### PR TITLE
GCE reacts to invalid cloud credentials.

### PR DESCRIPTION
**Pull request by anastasiamac**
## Description of change

This PR ensures that juju gce provider is responsive to google cloud un-happinness with provided credentials.

The most interesting bits are contained in ~/provider/gce/google/errors.go file. This is the code that examines google errors and determines whether the error is related to credential and worth reacting to.
All cloud calls that throw errors have now been modified to go through this logic to ensure that if the cloud complained about a credential, juju marks it as invalid.

Unit tests have been added to provide adequate initial coverage.

### CODE REVIEW

**Comment by anastasiamac** [comment]
[adjust gce provider to use callback to invalidate cloud credential when needed](https://trello.com/c/qR6MVUOh/349-adjust-gce-provider-to-use-callback-to-invalidate-cloud-credential-when-needed)

File: provider/gce/google/errors.go [hunk 1]
```diff
@@ -55,3 +60,57 @@ func NewMissingConfigValue(key, field string) error {
...
+				return true
+			}
+		}
+	}
+	return false
+
+}
+
+// AuthorisationFailureStatusCodes contains http status code nad description that signify authorisation difficulties.
+var AuthorisationFailureStatusCodes = map[int]string{
```
Comment (wallyworld) [inline provider/gce/google/errors.go]: I wonder if this is useful or needed in other providers? It seems quite generic.

Comment (anastasiamac) [reply provider/gce/google/errors.go]: Not really generic at all...

For one, for example, azure only has the codes not the description. Also the code for bad request is only applicable to oauth2-compliant clients. So for example, azure does not send it for authentication issues but gce does :)

So, at this stage, fairly provider specific. If I'll encounter duplication in other providers, will pull it into provider/common area or something...

File: provider/gce/disks.go [hunk 2]
```diff
@@ -462,27 +477,32 @@ func (v *volumeSource) attachOneVolume(volumeName string, mode google.DiskMode,
...
 	}
 	return attachment, nil
 }
 
 func (v *volumeSource) DetachVolumes(ctx context.ProviderCallContext, attachParams []storage.VolumeAttachmentParams) ([]error, error) {
 	result := make([]error, len(attachParams))
 	for i, volumeAttachment := range attachParams {
-		result[i] = v.detachOneVolume(volumeAttachment)
+		err := v.detachOneVolume(ctx, volumeAttachment)
+		if err != nil && google.HasDenialStatusCode(err) {
```
[the remark this answers is not in the document]
Comment (anastasiamac) [reply provider/gce/disks.go]: True :) and it does \o/ Thank you.

**Comment by anastasiamac** [comment]
$$merge$$

**Pull request merged**